addons/purchase/models/account_invoice.py

```diff
@@ -148,119 +149,312 @@ def write(self, vals):
+            # vendor bill.
+            matching_purchase_orders |= self.env['purchase.order'].search(
+                common_domain + [('name', 'in', po_references)])
+
+            if not matching_purchase_orders:
+                # If not found, we try looking for purchase orders whose `partner_ref` field matches one of the
+                # purchase order references in the vendor bill.
+                matching_purchase_orders |= self.env['purchase.order'].search(
+                    common_domain + [('partner_ref', 'in', po_references)])
+
+            if matching_purchase_orders:
+                # We found matching purchase orders and are extracting all purchase order lines together with their
+                # amounts still to be invoiced.
+                po_lines = [line for line in matching_purchase_orders.order_line if line.product_qty]
+                po_lines_with_amount = [{
                     'line': line,
                     'amount_to_invoice': (1 - line.qty_invoiced / line.product_qty) * line.price_total,
-                } for line in matching_pos.order_line if line.product_qty]
-
-                if amount_total - TOLERANCE < sum(line['amount_to_invoice'] for line in matching_pos_invoice_lines) < amount_total + TOLERANCE:
-                    return 'total_match', matching_pos.order_line
-
-                else:
-                    il_subset = self._find_matching_subset_invoice_lines(matching_pos_invoice_lines, amount_total, timeout)
-                    if il_subset:
-                        return 'subset_total_match', self.env['purchase.order.line'].union(*[line['line'] for line in il_subset])
+                } for line in po_lines]
+
+                # If the sum of all remaining amounts to be invoiced for these purchase orders' lines is within a
+                # tolerance from the vendor bill total, we have a total match. We return all purchase order lines
+                # summing up to this vendor bill's total (could be from multiple purchase orders).
+                if (amount_total - TOLERANCE
+                        < sum(line['amount_to_invoice'] for line in po_lines_with_amount)
+                        < amount_total + TOLERANCE):
+                    return ('total_match',
+                            matching_purchase_orders.order_line,
+                            None)
+
+                elif from_ocr:
+                    # The invoice comes from an OCR scan.
+                    # We try to match the invoice total with purchase order lines.
+                    matching_po_lines = self._find_matching_subset_po_lines(
+                        po_lines_with_amount, amount_total, timeout)
+                    if matching_po_lines:
+                        return ('subset_total_match',
+                                self.env['purchase.order.line'].union(
+                                    *[line for line in matching_po_lines]),
```

Comment:
what is this supposed to do? 
```
>>> amls = self.env['account.move.line'].browse([1, 2, 3])
>>> self.env['account.move.line'].union(*[x for x in amls])
account.move.line(1, 2, 3)
>>> self.env['account.move.line'].union(amls)
account.move.line(1, 2, 3)
>>> amls
account.move.line(1, 2, 3)
```

Reply:
The `matching_po_lines` is a Python `list` object, so it needs the `*` to pass it to `.union`. The `for` loop was not needed though (remaining from before when we took a specific item from `line`). So I removed the `for` loop, but kept the `*`.